javav2/example_code/inspector/src/main/java/com/java/inspector/InspectorActions.java

```diff
@@ -0,0 +1,829 @@
+                .severity(Collections.singletonList(severityFilter))
+                .build();
+
+        // Build filter creation request.
+        CreateFilterRequest request = CreateFilterRequest.builder()
+                .name(filterName)
+                .filterCriteria(filterCriteria)
+                .action(FilterAction.SUPPRESS)
+                .description(description)
+                .build();
+
+        return getAsyncClient().createFilter(request)
+                .whenComplete((response, exception) -> {
+                    if (exception != null) {
+
+                        Throwable cause = exception.getCause();
+
+                        if (cause instanceof Inspector2Exception) {
+                            Inspector2Exception e = (Inspector2Exception) cause;
+
+                            throw new CompletionException(
+                                    "Inspector2 service error: "
+                                            + e.awsErrorDetails().errorMessage(),
+                                    e
+                            );
+                        }
+
+                        // Unexpected async error
+                        throw new CompletionException(
+                                "Unexpected error creating filter: "
+                                        + exception.getMessage(),
+                                exception
+                        );
+                    }
+                })
+                // Extract and return the ARN of the created filter
+                .thenApply(response -> response.arn());
+    }
+    // snippet-end:[inspector.java2.create.filter.main]
+
+    // snippet-start:[inspector.java2.list_findings.main]
+
+    /**
+     * Fetches Inspector2 findings asynchronously using a paginator and returns
+     * a summary string when complete.
+     *
+     * @param maxResults     maximum findings per page
+     * @param filterCriteria optional filter, or null
+     * @return future completed with a summary of all collected findings
+     */
+    public CompletableFuture<String> listFindingsAsync(int maxResults,
+                                                       FilterCriteria filterCriteria) {
+        logger.info("Starting async paginator…");
+
+        final int PAGE_LIMIT = 4;
+        // Request builder
+        ListFindingsRequest.Builder builder =
+                ListFindingsRequest.builder()
+                        .maxResults(maxResults);
+
+        if (filterCriteria != null) {
+            builder.filterCriteria(filterCriteria);
+        }
+
+        ListFindingsRequest request = builder.build();
+        ListFindingsPublisher paginator =
+                getAsyncClient().listFindingsPaginator(request);
+
+        List<Finding> allFindings =
+                Collections.synchronizedList(new ArrayList<>());
+
+        CompletableFuture<String> future = new CompletableFuture<>();
+        paginator.subscribe(new Subscriber<ListFindingsResponse>() {
```

Comment:
The pagination in the latest updates looks way more complicated than necessary. It should follow the simplest pagination pattern, without page limits or overrides, like this:

`
```
List<JobSummary> jobSummaries = new ArrayList<>();
        ListJobsPublisher listJobsPaginator = getAsyncClient().listJobsPaginator(listJobsRequest);
        CompletableFuture<Void> future = listJobsPaginator.subscribe(response -> {
            jobSummaries.addAll(response.jobSummaryList());
        });
        future.join();
        return jobSummaries;
```
`

Reply:
For Inspector2 async paginators, you generally must use the reactive Subscriber pattern if you want fine-grained control

Other AWS SDKS -- like  S3 async paginators, the SDK provides subscribe() that returns a CompletableFuture, so you can use the simpler pattern.

Inspector2 async paginator → must use reactive Subscriber